### ISSUE TITLE
Measure what's wrong with our startup

### ISSUE THREAD

**Issue by heytherewill** [toggl/mobileapp#5954]
Our startup times aren't the best. We've been working on improving things, but we need to do more.

This issue is iOS specific, and in it we should list all things that can be improved (be they .NET specific or iOS specific) to make our startup as close to instantaneous as possible

To consider this issue closed we should have a write up on what can be improved and a list of follow-up issues to fix those things

**Comment by simonrozsival** [comment]
When measured in simulator, these parts of the initialization code have non-trivial durations:
```
EnsureInitialized - 80 ms
AppStart Initialization - 60 ms
GetAccessLevel - 180 ms
navigateAccordingToAccessLevel - 265 ms
LoginViewController.ViewDidLoad - 107 ms
```

The whole app start time when the user is not logged in is not very bad. On my iPhone 7 it is less than 2 seconds.

When logged in, there are these two additional non-trivial methods:
```
AppDelegate.loginWithCredentialsIfNecessary - 400 ms
AppDelegate.navigateAccordingToAccessLevel - 1250 ms
```

The start up time when logged in takes roughly 3.5 seconds on my iPhone 7.

**Comment by heytherewill** [comment]
Did you measure using a release build? Debug ones are really misleading in this sense. Also, the simulator data is really worthless for performance debugging purposes, so let's stick to actual device data.

If you drill down `navigateAccordingToAccessLevel`, what are the most problematic methods?
Same thing for `loginWithCredentialsIfNecessary`, 400ms is not ideal

**Comment by simonrozsival** [comment]
It was all debug builds. I'll try measuring it again tomorrow.

**Comment by simonrozsival** [comment]
I ran this on my iPhone 7 and this is what I got similar results:
```
// when logged in
EnsureInitialized - 77 ms
AppStart - 76 ms
GetAccessLevel - 150 ms
oginWithCredentialsIfNecessary - 140 ms
navigateAccordingToAccessLevel - 720 ms
```
Some of the times are better, but it's not very different from the simulator. I'll look into `navigateAccordingToAccessLevel` and find out what exactly is taking so long.

**Comment by simonrozsival** [comment]
I broke down the `navigateAccordingToAccessLevel`:
```
SyncManager.ForceFullSync - 186 ms -- this could happen in a different thread and we can shave off most of these 180 ms
Load MainTabBarViewModel - 231 ms
MainTabBarViewModel.Initialize - 184 ms
ViewControllerLocator.GetViewController - 127 ms
```

There's room for improvement. I noticed that the initialization of the VM actually means initializing all of the three tab VMs (log, reports, calendar). These should be initialized lazily in background threads I guess.

**Comment by simonrozsival** [comment]
The `loginWithCredentialsIfNecessary` method is much more interesting - the vast majority of the time spent by a single call inside `UserAccessManager.LoginWithSavedCredentials`: `userLoggedInSubject.OnNext(api);`

There are two places where we subscribe to `UserAccessManager.UserLoggedIn`:
- `BaseBackgroundSyncService.EnableBackgroundSync`
- `UIDependencyContainer.RecreateLazyDependenciesForLogin`

The `RecreateLazyDependenciesForLogin` method calls `WidgetsService?.Dispose();` which takes all the time (200 ms). This is something that must be improvable 🤔 I'll dig even deeper.

**Comment by simonrozsival** [comment]
The `Dispose` method simply disposes a subscription to the currently running time entry. It seems pointless to subscribe in the constructor of `WidgetsService` just to unsubscribe a few moments later. This must be fixed, but we must make sure we don't break some functionality of the Widgets project.

**Comment by simonrozsival** [comment]
I think this issue is now completed. I identified 3 problems with the app's startup:
1. Problem with incorrectly "recreating" widgets service - I already fixed this with #6618.
2. Problem with initiating full sync on startup - I created issue #6677 to address this.
3. Problem with initialization of all of the VMs in all of the tabs - I created issue #6679 to address this.

Is this enough or do you want more details, @heytherewill?

**Comment by heytherewill** [comment]
That should suffice for now :)
